File: ALittle/Module/ALittleIDE/Other/GameLibrary/Std/JSScript/Std/Loop/LoopList.js
{
if (typeof ALittle === "undefined") window.ALittle = {};


if (ALittle.LoopObject === undefined) throw new Error(" extends class:ALittle.LoopObject is undefined");
ALittle.LoopList = JavaScript.Class(ALittle.LoopObject, {
	Ctor : function() {
		this._count = 0;
		this._cur_index = 1;
		this._update_list = [];
		this._complete_callback = undefined;
	},
	get complete_callback() {
		return this._complete_callback;
	},
	set complete_callback(value) {
		this._complete_callback = value;
	},
	get total_count() {
		return this._count;
	},
	AddUpdater : function(value) {
		if (value === undefined) {
			return;
		}
		++ this._count;
		this._update_list[this._count - 1] = value;
	},
	RemoveUpdater : function(value) {
		if (value === undefined || this._count === 0) {
			return;
		}
		let ___OBJECT_1 = this._update_list;
		for (let index = 1; index <= ___OBJECT_1.length; ++index) {
			let updater = ___OBJECT_1[index - 1];
			if (updater === undefined) break;
			if (value === updater) {
				if (this._cur_index > index) {
					-- this._cur_index;
				}
				ALittle.List_Remove(this._update_list, index);
				-- this._count;
				break;
			}
		}
	},
	Reset : function() {
		this._cur_index = 1;
		let ___OBJECT_2 = this._update_list;
		for (let index = 1; index <= ___OBJECT_2.length; ++index) {
			let updater = ___OBJECT_2[index - 1];
			if (updater === undefined) break;
			updater.Reset();
		}
	},
	SetTime : function(time) {
		let ___OBJECT_3 = this._update_list;
		for (let index = 1; index <= ___OBJECT_3.length; ++index) {
			let updater = ___OBJECT_3[index - 1];
			if (updater === undefined) break;
			updater.SetTime(0);
		}
		this._cur_index = 1;
		if (time <= 0) {
			return [0, false];
		}
		let ___OBJECT_4 = this._update_list;
		for (let index = 1; index <= ___OBJECT_4.length; ++index) {
			let updater = ___OBJECT_4[index - 1];
			if (updater === undefined) break;
			let completed = undefined;
			[time, completed] = updater.SetTime(time);
			if (!completed) {
				this._cur_index = index;
				return [0, false];
			}
		}
		this._cur_index = this._count + 1;
		return [time, true];
	},
	IsCompleted : function() {
		return this._cur_index > this._count;
	},
	Completed : function() {
		if (this._complete_callback !== undefined) {
			this._complete_callback();
		}
	},
	SetCompleted : function() {
		for (let index = this._cur_index; index <= this._count; index += 1) {
			this._update_list[index - 1].SetCompleted();
		}
		this._cur_index = this._count + 1;
	},
	Update : function(frame_time) {
		if (this._cur_index > this._count) {
			return;
		}
		let updater = this._update_list[this._cur_index - 1];
		updater.Update(frame_time);
		if (updater.IsCompleted()) {
			++ this._cur_index;
			updater.Completed();
		}
	},
}, "ALittle.LoopList");

}

File: ALittle/Module/Std/Script/Std/Loop/LoopList.lua
-- ALittle Generate Lua
module("ALittle", package.seeall)

local ___thispath = select('1', ...):match("(.+[/\\]).+$") or ""
local ___rawset = rawset
local ___pairs = pairs
local ___ipairs = ipairs


assert(ALittle.LoopObject, " extends class:ALittle.LoopObject is nil")
LoopList = Lua.Class(ALittle.LoopObject, "ALittle.LoopList")

function LoopList:Ctor()
	___rawset(self, "_count", 0)
	___rawset(self, "_cur_index", 1)
	___rawset(self, "_update_list", {})
	___rawset(self, "_complete_callback", nil)
end

function LoopList.__getter:complete_callback()
	return self._complete_callback
end

function LoopList.__setter:complete_callback(value)
	self._complete_callback = value
end

function LoopList.__getter:total_count()
	return self._count
end

function LoopList:AddUpdater(value)
	if value == nil then
		return
	end
	self._count = self._count + 1
	self._update_list[self._count] = value
end

function LoopList:RemoveUpdater(value)
	if value == nil or self._count == 0 then
		return
	end
	for index, updater in ___ipairs(self._update_list) do
		if value == updater then
			if self._cur_index > index then
				self._cur_index = self._cur_index - 1
			end
			List_Remove(self._update_list, index)
			self._count = self._count - 1
			break
		end
	end
end

function LoopList:Reset()
	self._cur_index = 1
	for index, updater in ___ipairs(self._update_list) do
		updater:Reset()
	end
end

function LoopList:SetTime(time)
	for index, updater in ___ipairs(self._update_list) do
		updater:SetTime(0)
	end
	self._cur_index = 1
	if time <= 0 then
		return 0, false
	end
	for index, updater in ___ipairs(self._update_list) do
		local completed
		time, completed = updater:SetTime(time)
		if not completed then
			self._cur_index = index
			return 0, false
		end
	end
	self._cur_index = self._count + 1
	return time, true
end

function LoopList:IsCompleted()
	return self._cur_index > self._count
end

function LoopList:Completed()
	if self._complete_callback ~= nil then
		self._complete_callback()
	end
end

function LoopList:SetCompleted()
	local index = self._cur_index
	while true do
		if not(index <= self._count) then break end
		self._update_list[index]:SetCompleted()
		index = index+(1)
	end
	self._cur_index = self._count + 1
end

function LoopList:Update(frame_time)
	if self._cur_index > self._count then
		return
	end
	local updater = self._update_list[self._cur_index]
	updater:Update(frame_time)
	if updater:IsCompleted() then
		self._cur_index = self._cur_index + 1
		updater:Completed()
	end
end

